### ISSUE TITLE
Fix S3655  FP: [DoesNotReturn] is not respected

### ISSUE THREAD

**Issue by mu88** [SonarSource/sonar-dotnet#8633]
### Description

The issue `Empty nullable value should not be accessed - csharpsquid:S3655` is raised when accessing a nullable value type (e. g. `TimeOnly?`) although a custom `throw` helper (annotated with `[DoesNotReturn]`) ensures that it cannot be `null`.

### Repro steps

Please see the following SSCCE:
```csharp
using System.Diagnostics.CodeAnalysis;

namespace ReproSonarDoesNotReturn;

public class Class1
{
    public TimeOnly GetCurrentTime()
    {
        var currentTime = LoadCurrentTimeFromProvider();
        if (!currentTime.HasValue)
        {
            ExceptionHelper.ThrowBecauseOfMissingTime();
        }

        return currentTime.Value;
    }

    private TimeOnly? LoadCurrentTimeFromProvider() => null; // shortened for demo purposes
}

public static class ExceptionHelper
{
    [DoesNotReturn]
    public static void ThrowBecauseOfMissingTime() => throw new ArgumentNullException();
}
```

Here are the necessary steps to reproduce the issue:
1. Run `dotnet new tool-manifest`
2. Run `dotnet tool install dotnet-sonarscanner --version 6.1.0`
3. Run `dotnet dotnet-sonarscanner begin /k:repro-sonar-does-not-return /d:sonar.host.url=<<your Sonar instance>>`
4. Run `dotnet build`

### Expected behavior

The build output should not raise the Sonar issue `csharpsquid:S3655`.

### Actual behavior

The output contains the following issues:
```shell
MSBuild version 17.8.3+195e7f5a3 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\work\temp\ReproSonarDoesNotReturn\Class1.cs(24,61): warning S3928: Use a constructor overloads that allows a more meaningful exception message to be provided. (https://rules.sonarsource.com/csharp/RSPEC-3928) 
[C:\work\temp\ReproSonarDoesNotReturn\ReproSonarDoesNotReturn.csproj]
C:\work\temp\ReproSonarDoesNotReturn\Class1.cs(15,16): warning S3655: 'currentTime' is null on at least one execution path. (https://rules.sonarsource.com/csharp/RSPEC-3655) [C:\work\temp\ReproSonarDoesNotReturn\ 
ReproSonarDoesNotReturn.csproj]
  ReproSonarDoesNotReturn -> C:\work\temp\ReproSonarDoesNotReturn\bin\Debug\net8.0\ReproSonarDoesNotReturn.dll
  Sonar: (ReproSonarDoesNotReturn.csproj) Project processed successfully

Build succeeded.

C:\work\temp\ReproSonarDoesNotReturn\Class1.cs(24,61): warning S3928: Use a constructor overloads that allows a more meaningful exception message to be provided. (https://rules.sonarsource.com/csharp/RSPEC-3928)  
[C:\work\temp\ReproSonarDoesNotReturn\ReproSonarDoesNotReturn.csproj]
C:\work\temp\ReproSonarDoesNotReturn\Class1.cs(15,16): warning S3655: 'currentTime' is null on at least one execution path. (https://rules.sonarsource.com/csharp/RSPEC-3655) [C:\work\temp\ReproSonarDoesNotReturn\ 
ReproSonarDoesNotReturn.csproj]
    2 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.07
```

### Known workarounds

There are no proper workarounds known yet, except disabling the issue, or adding redundant code (`currentTime.HasValue`) - but I wouldn't consider them workarounds.

### Related information

* C# 12
* Rider 2023.3.3
* MSBuild `17.8.3+195e7f5a3`, .NET `8.0.101`
* SonarScanner for .NET version (if used) `6.1.0`
* Windows 10

**Comment by martin-strecker-sonarsource** [comment]
Hello @mu88 

thank you for reporting this issue. We can re-produce this on our side. See #8637.
The issue is only reproducible in our "old" implementation of the rule, which was replaced by a new implementation in our 8.56 release from April 2023:

https://github.com/SonarSource/sonar-dotnet/releases/tag/8.56.0.67649

You probably use an older version of the analyzer. You can see the analyzer version in the begin step output. The output looks like so:
```
dotnet dotnet-sonarscanner begin /k:repro-sonar-does-not-return /d:sonar.host.url=<<your Sonar instance>>
...
Processing plugin: csharp version 8.55.0.65544
...
```

You can upgrade to a newer version of the analyzer by updating your SonarQube instance.

I'm closing the issue as "fixed". If you use a more recent version of the plugin than 8.56, please leave a comment and I will re-open it.
